### PR TITLE
FirmwareUploder Process with StreamGobblerThreads exemple.

### PR DESCRIPTION
**Pull request by PPAC37**
From an old project of my ( to review but working)

Only for non interactive commands.

The key is in the StreamGobbler (threaded object) as "process err stream" can be interlaced with  "process out stream".

To adatapt to the needs.

### CODE REVIEW

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 1]
```diff
@@ -47,25 +82,40 @@ public void run(String hexPath,String portName) throws Exception {
 			Log.message("Trying 2 "+(p.resolve("../../etc/avrdude.conf").toString()));
 			f = p.resolve("../../etc/avrdude.conf").toFile();
 			if(!f.exists()) {
-				throw new Exception("Cannot find nearby avrdude.conf");
+			    // TODO to reactive the throw juste for demo purpose of process exec.
+				//throw new Exception("Cannot find nearby avrdude.conf");
+				   System.err.println("Cannot find nearby avrdude.conf ");
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:87]: please use `Log.error` or later if #402 `logger.error`

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 2]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
 		} catch(Exception e) {
 			e.printStackTrace();
 		}
 	}
+	
+	//
+	//
+	//
+	
+	public static SimpleDateFormat simpleDateFormat = new SimpleDateFormat("ddMMyyyy hh:mm:ss");
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:187]: SimpleDateFormat is not thread safe thus must be reinstancited each you need it

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 3]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+	}
+	// TO REVIEW / TODO le cas d'un process interactif ou sans fin ...
+	// TO REVEIW ( as this is a way to exec commands on the user system ... )
+	// should be a least private for security (and all methodes that using it ) ?
+	protected static String execBashCommand(String[] cmdArray, StreamGobblerReadLineBufferedSpecial streamGobblerProcessIn,SelectTextArea  execResult) {
+	    lastExecSucces = false;
+	    final StringBuffer res = new StringBuffer();
+	    res.setLength(0);
+	    if (debugRuntimeExecPre) {
+		System.out.printf("Running : ");
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:211]: `Log.message()` for every message

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 4]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+		for (String arg : cmdArray) {
+		    System.out.printf("%s ", arg);
+		}
+		System.out.printf("\n");
+	    }
+	    res.append("Running : ");
+	    for (String arg : cmdArray) {
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:218]: duplicate from above ?

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 5]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+		res.append(String.format("%s ", arg));
+	    }
+	    res.append("\n");
+
+	    try {
+		Date dStar = new Date();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:224]: indentation : in your IDE, you should use the format feature.

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 6]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+		//pb.directory(new File("myDir"));
+
+		// To merge err on out // TO REVIEW for specials case.
+		pb = pb.redirectErrorStream(true);
+		
+		//Process 
+		process = pb.start();
+
+		if (streamGobblerProcessIn == null) {
+		    streamGobblerProcessIn = new StreamGobblerReadLineBufferedSpecial(process.getInputStream(), "out") {
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:240]: I think the flash process should not ask anything to the user thus capturing the inputstream seems not necessary, what do you think ?

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 7]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+		streamGobblerProcessIn.join();
+		
+		
+		// Wait for the process to end.
+		process.waitFor();
+		int ret = process.exitValue();
+		Date dEnd = new Date();
+
+		if (ret == 0) {
+		    lastExecSucces = true;
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:340]:
```suggestion
		    lastExecSucces = ret == 0;
```

File: src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java [hunk 8]
```diff
@@ -120,13 +170,203 @@ public void setAvrdudePath(String avrdudePath) {
...
+		
+		if (debugRuntimeExecPre) {
+		    System.out.printf("Running : ");
+		    for (String arg : cmdArray) {
+			System.out.printf("%s ", arg);
+		    }
+		    System.out.printf("\nexit on error = %s\n", e.getMessage());
+		}
+		
+		res.append(e.getMessage());
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/firmwareUploader/FirmwareUploader.java:366]: The full stack trace is missing

**Review by coliss86 — CHANGES_REQUESTED** [review]
Good job !
The indentation of the code is not correct, make it hard to read.
A lot of comments indicate that it is not finished and it cannot be merged. You should mark the PR as Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by PPAC37** [comment]
Thank you.

I close this PR to make a draft PR immediately!

I take note of all the comments. My apologies if I do not correct this immediately. (I take advantage of the holiday season with the family.)

And maybe i should wait for you PR https://github.com/MarginallyClever/Makelangelo-software/pull/402  to be finish and  merged.

Good continuation, and happy holidays.

**Comment by coliss86** [comment]
No don't close :)
You can change the status of this PR (I think) or at least, you can edit the title to "DRAFT:"

**Comment by PPAC37** [comment]
Sorry I closed the PR!
 (I'm starting on github and haven't figured out how to pass an open PR into a PR project yet ... I only found the solution to close it and make one new but this time in draft. https://github.com/MarginallyClever/Makelangelo-software/pull/406 )

**Pull request closed without merging**